pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -502,6 +502,13 @@ func (r *ReconcileClusterSync) Reconcile(ctx context.Context, request reconcile.
 	}
 	// Clear the cluster sync failing metric
 	if !failing {
+		// Set the metric to zero first since clearing the metric would only stop recording it.
+		cond := controllerutils.FindClusterSyncCondition(origStatus.Conditions, hiveintv1alpha1.ClusterSyncFailed)
+		if cond != nil && hivemetrics.ShouldLogGaugeDurationMetric(hivemetrics.MetricClusterSyncFailingSeconds,
+			time.Since(cond.LastTransitionTime.Time).Seconds()) {
+			hivemetrics.MetricClusterSyncFailingSeconds.WithLabelValues(
+				hivemetrics.GetNameSpacedName(clusterSync.Namespace, clusterSync.Name)).Set(0)
```

Comment:
I think we also need to zero the metric in [this code path](https://github.com/openshift/hive/blob/e7fbba49305cab53e8e70ec433ebb3e9ac6e7cae/pkg/controller/metrics/metrics.go#L484-L489) by adding an `else` branch.

Having looked at that, I'm actually thinking we may want to get rid of handling the metric in this controller at all -- remove this whole block. Thoughts?

Reply:
`ShouldLogXDurationMetric` requires seconds as a parameter - seconds since the `lastTransitionTime`. While the clusterSync is still failing, we can make that decision from the actual status, but when we need to clear the metric and set it to zero - that's when we need the lastTransitionTime of the condition before we change it - that is only available in the clusterSync controller. Including `else` part where the metrics controller computes the metric and log it to zero regardless of whether we logged the metric for the CS or not - needlessly complicates the logic and ends up with more entries than necessary
Clearing the metric is a different case because it clears it whether or not the metric was ever logged
Besides, the test for WaitingForCO metric [shows](https://user-images.githubusercontent.com/54181650/180893909-66c85790-e42d-4b4e-927e-7eaba8dfce68.png) how the metrics controller eventually does stop logging the metric, so I do not see this as a deal breaker

Reply:
Oh. This is a good point. If sync succeeds, setFailedCondition (L487) will set the ClusterSyncFailed condition to False and update the LastTransitionTime to Now(). So ShouldLogGaugeDurationMetric won't trigger (unless configured to zero, which shouldn't be the case), and this branch won't be reached.

Point is, we're going to have to solve this problem regardless.

Since we're deleting it immediately, wouldn't it be okay to set it to zero even if it doesn't exist beforehand? It would take rather extreme timing for it to get scraped in that interval -- and I'm willing to accept that it would be zero in that case.

But wait again. Isn't setting to zero a no-op anyway if we're then going to delete right away? The scraper won't see that zero; it'll just see that the metric isn't there, and use the previously-reported value -- which is the whole problem we're trying to solve with this PR.

Let's sleep on this and discuss tomorrow.